### ISSUE TITLE
Convert markdown in Yaml fields to JSX

### ISSUE THREAD

**Issue by vai0** [gatsbyjs/gatsby#21789]
Our site uses Gatsby with NetlifyCMS writing content into .yml files. There are many yml files with field values that contain markdown, with values arbitrarily nested in objects and lists. For example,

```yml
# solutions.yml
overviewSection: >-
  # h1
  # h2
  [some link](https://www.google.com)
faqSection:
  overview: >-
    more __markdown__
  blocks:
    - title: non-markdown string
        copy: >-
          # h1
          # h2
          [some link](https://www.google.com)
```

We're currently converting the markdown to HTML, then injecting it via `dangerouslySetInnerHTML`. This is not ideal since we now need to maintain separate HTML styles that visually map 1-1 to our existing typography components.

My goal is to convert the markdown strings to compiled JSX on build so it can be queried and passed into `<MDXRenderer />`.   

e.g., 
```js
query {
  pagesYaml {
    overviewSection // <-- compiled JSX
    faqSection {
      overview // <- compiled JSX
      blocks {
        title
        copy // <-- compiled JSX
      }
    }
  }
}
```
### 1. Use [`gatsby-transformer-yaml-full`](https://github.com/stldo/gatsby-transformer-yaml-full/tree/master/packages/gatsby-transformer-yaml-full#readme) which receives a set plugins that allow you to custom parse yml values - create a plugin which looks for the yaml tag `!mdx` and converts the value to compiled JSX.

This is the only solution I've found that really works

Given the yml file is...
```yml
# solutions.yml
faqSection:
  blocks:
    - title: non-markdown string
      copy: !mdx |
        # h1
        # h2
        [some link](https://www.google.com)
```

The graphql query
```js
query {
  pagesYaml {
    faqSection {
      blocks {
        title
        copy
      }
    }
  }
}
```
returns `copy` as compiled JSX
```json
{
  "data": {
    "pagesYaml": {
      "faqSection": {
        "blocks": [
          {
            "title": "non-markdown string",
            "copy": "function _extends() { _extends = Object.assign ... MDXContent.isMDXComponent = true;"
          }
        ]
      }
    }
  }
}
```
`copy` is then fed into `gatsby-plugin-mdx`'s `<MDXRenderer />`.
```js
// pages/solutions.js
import React from "react";
import { MDXRenderer } from "gatsby-plugin-mdx";
...
export default ({ data }) => (
  ...
  {data.pagesYaml.faqSection.blocks.map(({ copy }) => (
    <MDXRenderer>{copy}</MDXRenderer>
  ))}
  ...
)

```

The caveat to this is the custom plugin - I'm lifting internals from `gatsby-plugin-mdx`, specifically the babel transforms that generates `Mdx.body` node field.
```js
// plugins/mdx-yaml-full/index.js

const mdx = require("@mdx-js/mdx")
const babel = require("@babel/core")
const BabelPluginPluckImports = require("gatsby-plugin-mdx/utils/babel-plugin-pluck-imports")
const htmlAttrToJSXAttr = require("gatsby-plugin-mdx/utils/babel-plugin-html-attr-to-jsx-attr")
const removeExportKeywords = require("gatsby-plugin-mdx/utils/babel-plugin-remove-export-keywords")

module.exports = ({ node }, pluginOptions) => (
  tag: "!mdx",
  options: {
    kind: "scalar",
    construct: async data => {
      const code = await mdx(data)

      // lifted straight from this section:
      // https://github.com/gatsbyjs/gatsby/blob/cd150b5e5264a5a75f2abc27e3430c55bfcd4e41/packages/gatsby-plugin-mdx/utils/gen-mdx.js#L138
      const transformed = babel
        .transformSync(code, {
          plugins: [
            new BabelPluginPluckImports().plugin,
            htmlAttrToJSXAttr,
            removeExportKeywords,
          ],
          presets: [
            require("@babel/preset-react"),
            [
              require(`@babel/preset-env`),
              {
                useBuiltIns: `entry`,
                corejs: 2,
                modules: `false`,
              },
            ],
          ],
        }) 
        // ... ew :(
        .code.replace(
          /export\s*default\s*function\s*MDXContent\s*/,
          `return function MDXContent`
        )
        .replace(
          /export\s*{\s*MDXContent\s+as\s+default\s*};?/,
          `return MDXContent;`
        )

      return transformed
    },
  },
)

```
<br /><br />
### 2. Extend the pagesYaml node type with [`exports.createResolvers`](https://www.gatsbyjs.org/docs/schema-customization/#extending-third-party-types)

and add an additional field alongside the markdown content, converted to JSX. If I understand how the this works though - this would again very quickly be cumbersome because every single field would need to be mapped into `createResolvers`.

```js
const MDXToJSX = require("mdx-to-jsx");

// gatsby-node.js
exports.createResolvers = ({
  actions,
  cache,
  createNodeId,
  createResolvers,
  store,
  reporter,
}) => {
  const { createNode } = actions
  createResolvers({
    pagesYaml: {
      // somehow target the "copy" field nested in "faqSection.blocks[].copy" 
      copy: {
        type: `StringToMdx`,
        resolve(source, args, context, info) {
          return createRemoteFileNode({
            copy: MDXToJSX(source.copy),
            store,
            cache,
            createNode,
            createNodeId,
            reporter,
          })
        },
      },
      // and more keys per field we want compiled...
    },
  })
}
```
<br /><br />
### 3. Create nodeFields attached to the node
which was mentioned as a solution here: https://github.com/gatsbyjs/gatsby/issues/19119

In `onCreateNode`, create an Mdx node when a node field appended with `__mdx` is found and attach the value as a `nodeField` to `pagesYaml`. The big downside to this approach is because `nodeFields` only takes scalar values - a node field needs to be created for each and every markdown field - which quickly becomes verrrry cumbersome for the client to query and re-parse down the road. e.g.,

Given the yml file looks like...
```yml
# solutions.yml
faqSection:
  blocks:
    - title: non-markdown syntax string
      copy__mdx:
        # h1
        # h2
        [some link](https://www.google.com)
```
```js
exports.onCreateNode({ node, actions }) => async {
  const { createNode, createNodeField } = actions;
  
  // Find node.keys appended with __mdx. e.g., copy__mdx, and return
  // the dot path/value as key/value
  // {
  //     "faqSection.blocks[0].copy__mdx": "# h1 #h2 [some link](https://www.google.com)",
  //     ...
  // }
  const mdxFields = getMdxFields(node);

  if (mdxFields) {
    _.each(mdxFields, (content, dotPath) => {
      createNode({
        type: "StringToMdx",
        id: `${node.id} >>> Mdx`
        ...
        internals: {
          mediaType: "text/markdown",
          content
          ...
        }
      })

      // converts faqSection.blocks[0].copy__mdx to faqSection_blocks_0_copy__mdx
      const safeDotPath = safeDotPath(dotPath);

      createNodeField({
        node,
        name: `${safeDotPath}___NODE`,
        value: node.id
      })
    })
  }
    ...
}
```
which results in an unwieldy query and the extra work to merge non-mdx and mdx fields back together client-side.
```js
query {
  pagesYaml {
    faqSection {
      blocks {
        title
      }
    }
    fields {
      faqSection_blocks_0_copy
      faqSection_blocks_1_copy
      faqSection_blocks_2_copy
      ...
    }
  }
}
```

**To summarize, is there a suggested method to compile markdown field values that are arbitrarily nested to JSX?**
Has anyone else encountered and found a workaround to this problem?

**Comment by pieh** [comment]
Would something like this work better:
https://github.com/pieh/i21789/blob/master/gatsby-node.js

it still require to explicitly define which fields get the treatment, but it's using `gatsby-plugin-mdx` and (hopefully) easier to follow setup.

There seems to be some weirdnes with newlines in data from yaml, but this actual data we get from yaml transformer (missing newlines, everything is concataneted - not sure if a bug, or this is just how yaml supposed to work):
![Screenshot 2020-02-27 at 19 31 53](https://user-images.githubusercontent.com/419821/75474786-e5e2c800-5997-11ea-8ab7-4de9eea39cad.png)

In any case result is:
![Screenshot 2020-02-27 at 19 33 08](https://user-images.githubusercontent.com/419821/75474848-00b53c80-5998-11ea-9ff7-75cb0a6a5303.png)

(that weird lack of newlines shown here):
![Screenshot 2020-02-27 at 19 33 30](https://user-images.githubusercontent.com/419821/75474880-0dd22b80-5998-11ea-803f-db2eceac9c8a.png)
(using https://github.com/pieh/i21789/blob/master/src/pages/index.js )

**Comment by vai0** [comment]
@pieh Thank you for the quick response! Did not think about the approach with field extensions. Although the need to explicitly define the types is what we're trying to avoid...  it'd get a bit hairy when we have ~40 or so yml files and increasing, with probably 2-3 types per file to define.

Would there be another approach to implement this while using `gatsby-plugin-mdx` without having to explicitly define the types?

**Comment by pieh** [comment]
Question - would every field be considered as mdx field or not? It might be possible to do if every field would be (tho would need to change my sample drastically). 

The point here is that we need to define resolver types not based on data, but rather on only field names. We do need to at least declare if they are Strings or not.

We can then do similar thing as `gatsby-transformer-yaml-full` and use `!mdx |` prefix to figure out if it should be "mdx code" or just regular string

**Comment by pieh** [comment]
Alternatively - we could create extra `${fieldName}Mdx` fields for each existing field and this is how you would chose in your queries if you want mdx code or direct value

**Comment by vai0** [comment]
> Question - would every field be considered as mdx field or not? It might be possible to do if every field would be (tho would need to change my sample drastically).
Not all fields would be considered an mdx field. We have some field values across some of the Yaml files which are number-based. Although we can treat them to strings if it makes implementation cleaner.    

> The point here is that we need to define resolver types not based on data, but rather on only field names. We do need to at least declare if they are Strings or not.
> 
> We can then do similar thing as gatsby-transformer-yaml-full and use !mdx | prefix to figure out if it should be "mdx code" or just regular string
Can you clarify?

> Alternatively - we could create extra ${fieldName}Mdx fields for each existing field and this is how you would chose in your queries if you want mdx code or direct value
hmm, this could work if it doesn't affect build times. How would something like this look?

**Comment by zslabs** [comment]
This sounds like a really interesting topic that I think would be useful to go along with https://github.com/gatsbyjs/gatsby/issues/18570
